state/store/defaultstore/message.go

```diff
@@ -73,68 +73,129 @@ func (s *Message) MaxMessages() int {
+func (s *Message) MessageSet(message discord.Message, update bool) error {
+	if s.maxMsgs <= 0 {
+		return nil
+	}
+
 	iv, _ := s.channels.LoadOrStore(message.ChannelID)
 
 	msgs := iv.(*messages)
 
 	msgs.mut.Lock()
 	defer msgs.mut.Unlock()
 
-	for i, m := range msgs.messages {
-		if m.ID == message.ID {
-			DiffMessage(message, &m)
-			msgs.messages[i] = m
-			return nil
+	if update {
+		// Opt for a linear latest-to-oldest search in favor of something like
+		// sort.Search, since more recent messages are more likely to be edited
+		// than older ones.
+		for i, oldMessage := range msgs.messages {
+			// We found a match, update it.
+			if oldMessage.ID == message.ID {
+				DiffMessage(message, &oldMessage)
+				msgs.messages[i] = oldMessage // Now updated.
+				return nil
+			}
 		}
+
+		return nil
 	}
 
-	// Order: latest to earliest, similar to the API.
+	switch {
+	case len(msgs.messages) == 0:
+		msgs.messages = []discord.Message{message}
+	case shouldPrependMessage(message, msgs.messages):
+		if len(msgs.messages) == s.maxMsgs {
+			copy(msgs.messages[1:], msgs.messages)
+			msgs.messages[0] = message
+		} else {
+			msgs.messages = append([]discord.Message{message}, msgs.messages...)
+		}
+	case shouldAppendMessage(message, msgs.messages) && len(msgs.messages) < s.maxMsgs:
+		msgs.messages = append(msgs.messages, message)
+	}
 
-	// Check if we already have the message. Try to derive the order otherwise.
-	var insertAt int
+	// We already have this message or we can't append any more messages.
+	return nil
+}
 
-	// Since we make the order guarantee ourselves, we can trust that we're
-	// iterating from latest to earliest.
-	for insertAt < len(msgs.messages) {
-		// Check if the new message is older. If it is, then we should insert it
-		// right after this message (or before this message in the list; i-1).
-		if message.ID > msgs.messages[insertAt].ID {
-			break
+// shouldPrependMessage checks if the passed message may be prepended to the
+// passed message slice, that is ordered from latest to oldest.
+//
+// shouldPrependMessage is biased as it will return true if the timestamps
+// of the passed message and the latest message match, even though the true
+// order cannot be determined in that case.
+func shouldPrependMessage(message discord.Message, messages []discord.Message) bool {
+	// The id of message is greater than the one of the first aka. newest
+	// message. It is therefore younger, and should be inserted.
+	if message.ID > messages[0].ID {
+		return true
+	}
+
+	// Two cases remain, the timestamps are equal or message is older than our
+	// first message. So we compare timestamps. If they are equal, make sure
+	// messages doesn't contain a message with the same id, in order to prevent
+	// insertion of a duplicate.
+	// ID timestamps are used, as they provide millisecond accuracy in contrast
+	// to the second accuracy of discord.Message.Timestamp.
+	ts := message.ID >> 22
```

Comment:
The timestamp can be obtained with `message.ID.Time()`.

Reply:
I know, this syntax was just shorted to compare, and (subjectively) more readable: `ts < message[i].ID>>22` vs. `ts.Equal(message[i].ID.Time())`.

Reply:
I think it's better to change it, since `>>22` doesn't seem immediately obvious.